daml-lf/tests/MultiKeys.daml

```diff
@@ -0,0 +1,232 @@
+      controller p
+      do _ <- fetchByKey @Keyed p
+         cid1 <- create (Keyed p)
+         (cid, _) <- fetchByKey @Keyed p
+         cid1 === cid
+
+    -- should be accepted in uck, cid not fetched by key
+    nonconsuming choice FetchDoesNotOverwriteGlobal : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do (cidByKey, _) <- fetchByKey @Keyed p
+         cid =/= cidByKey
+         c <- fetch cid
+         key c === p
+         (cidByKey', _) <- fetchByKey @Keyed p
+         cidByKey' === cidByKey
+
+    -- should be accepted in uck, cid not fetched by key
+    nonconsuming choice FetchDoesNotOverwriteLocal : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do local <- create (Keyed p)
+         c <- fetch cid
+         key c === p
+         (cid', _) <- fetchByKey @Keyed p
+         local === cid'
+
+    -- should be accepted in uck
+    nonconsuming choice LocalArchiveOverwritesUnknownGlobal : ()
+      controller p
+      do -- assume that a global contract with same key exists but don’t fetch it.
+         local <- create (Keyed p)
+         archive local
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be rejected in uck
+    nonconsuming choice LocalArchiveOverwritesKnownGlobal : ()
+      controller p
+      do _ <- fetchByKey @Keyed p
+         local <- create (Keyed p)
+         archive local
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- Should be accepted in uck
+    nonconsuming choice GlobalArchiveOverwritesUnknownGlobal : ()
+      with
+        cid1 : ContractId Keyed
+        cid2 : ContractId Keyed
+      controller p
+      do cid1 =/= cid2
+         c1 <- fetch cid1
+         key c1 === p
+         c2 <- fetch cid2
+         key c2 === p
+         archive cid1
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be accepted in uck
+    nonconsuming choice GlobalArchiveOverwritesKnownGlobal1 : ()
+      with
+        cid1 : ContractId Keyed
+        cid2 : ContractId Keyed
+      controller p
+      do cid1 =/= cid2
+         c1 <- fetch cid1
+         key c1 === p
+         c2 <- fetch cid2
+         key c2 === p
+         (keyCid, _) <- fetchByKey @Keyed p
+         assert (keyCid `elem` [cid1, cid2])
+         -- A consuming choice is
+         -- implicitly a by-key operation.
+         archive (fromSome (find (/= keyCid) [cid1, cid2]))
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be accepted in uck
+    nonconsuming choice GlobalArchiveOverwritesKnownGlobal2 : ()
+      with
+        cid1 : ContractId Keyed
+        cid2 : ContractId Keyed
+      controller p
+      do cid1 =/= cid2
+         c1 <- fetch cid1
+         key c1 === p
+         c2 <- fetch cid2
+         key c2 === p
+         (keyCid, _) <- fetchByKey @Keyed p
+         assert (keyCid `elem` [cid1, cid2])
+         archive keyCid
+         None <- lookupByKey @Keyed p
```

Comment:
I agree that `None` is the only sensible choice here. `fetchByKey` brought the association `key -> Some keyCid` into scope, so it's `key -> None` after the archival.

Reply:
I'm not sure for non-uck mode. The other option would be to query the ledger again for another contract. See my comment above: https://github.com/digital-asset/daml/pull/9472/files#r620200392

Reply:
If there's a chance that `lookupByKey` returns a different contract here, then we need to change the Canton implementation of contract keys, because Canton currently supports at most one input contract for a contract key.

Reply:
As discussed, we’re postponing multikey semantics so resolving this comment for this PR.